[ao][sparsity] Implementation of squash_mask()

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* __->__ #80889
* #80888
* #80887
* #80886

Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

Commit:
[ao][sparsity] Implementation of squash_mask()

Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -270,10 +270,34 @@ def update_mask(self, name, data, configs):
+                return input_data
+        return hook
+
     def squash_mask(self, **kwargs):
```
Aren't you supposed to deregister the hooks at this point? Does `.remove` do that?

Also, I think we should ask the user if they want to keep the forward pre hook. I am just thinking, if the user just want the masks, then the squash_mask is supposed to compute it, and get out, right? So a user can *optionally* request the forward hook to stay

macandro96 replied:
yes ```.remove()``` deregisters the hook as mentioned in [docs](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_pre_hook.html)

Also, ```squash_mask()``` does not compute the mask, ```update_mask()``` does. We can add an argument ```attach_sparsify_hook``` which if True attache the sparsify forward pre-hook, otherwise just removes the aggregate handle

jerryzh168 commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -270,10 +270,34 @@ def update_mask(self, name, data, configs):
                 data_feature = reduce_fn(data[feature_idx])
                 mask[feature_idx].data = mask_fn(data_feature, **sparse_config)
 
+    def __sparsify_hook(self, name):
```
nit: rename to `sparsify_pre_hook` to make it clearer?

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

z-a-f reviewed (approved):
@macandro96 Check the tests + dynamo. Once fixed, you can merge

Commit:
Update on "[ao][sparsity] Implementation of squash_mask()"


Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
